SCRD-3501 Update to add compute node in replace details

guglo opened the pull request:
This commit includes:

1. Added section to have inputs for new compute node in repalce server
   details modal. Replace compute node will add compute node first.

2. Fixed issue when replace non-compute node saves the replace
   information to the model.

Commit:
SCRD-3501 Update to add compute node in replace details

This commit includes:

1. Added section to have inputs for new compute node in repalce server
   details modal. Replace compute node will add compute node first.

2. Fixed issue when replace non-compute node saves the replace
   information to the model.

Commit:
Fixed a couple issues:

1. Only update mac, impi info for replace non-compute host
2. Save replacement host to saved servers.
3. valid mac and impi info when they have inputs
4. Added a little more bottom margin for wipe data disks option so
it won't be too crowded when everything expanded in the dialog

Commit:
Minor cleanup

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -66,22 +65,81 @@ class ReplaceServerDetails extends Component {
       (this.state.isOsInstallPasswordValid || !this.state.isInstallOsSelected);
   }
 
   handleDone = () => {
```
The function has some problems with naming and cohesion that I am really struggling to understand, and now that we are making substantive changes to this, I'd like to see it fixed if possible:
1. The name `data` is too generic to make any sense of, and it needs a better name. It it being passed to a function in the caller that is expecting a parameter of `server`, so I think server would be a much better name here.
2. `theProps` is similarly so generic that it is impossible to make sense out of. Worse yet, the caller seems to be using it in a variety of ways -- one function is looking for a `wipeDisk` field and ignoring the rest, while other code is doing other things. It is really difficult to understand how this is going to be populated, and it is not clear why all of these fields have to be combined together in ReplaceServerDetails into one big object. It would be far more easy to understand to call `this.props.doneAction` and explicitly list each parameter instead of combining them, i.e.
this.props.doneAction(server, this.state.isWipeDiskSelected, this.state.isInstallOsSelected, ...)
and then let the caller, `UpdateServers`, use them or package them according to its needs.

guglo replied:
will update

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -92,19 +150,26 @@ class ReplaceServerDetails extends Component {
-      // generate a new uid
-      data.uid = genUID();
+      // generate a new uid, treat it as manual added server
+      data['uid'] = genUID('manul');
```
This dialog box should just be returning what the selectedServerId is, and the caller should be generatign a new ID if one is necessary. (BTW `manual` is misspelled here)

guglo replied:
will update

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -92,19 +150,26 @@ class ReplaceServerDetails extends Component {
+
+    // will record the old compute host information so we can process removing
+    // the compute host later
+    if(this.isCompute()) {
```
This does not look right. The caller is the one supplying these props values, so we should not be returning them back to the caller. The caller can build up this structure if needed.

guglo replied:
will update

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -66,22 +65,81 @@ class ReplaceServerDetails extends Component {
+      MODEL_SERVER_PROPS.forEach(input_name => {
+        data[input_name] = this.state.inputValue.get(input_name);
+      });
+      data['role'] = this.props.data['role'];
```
This does not look right. The caller is supplying this data already, so we should not be turning around and feeding it right back. Let the caller insert it into this object if it is needed.

guglo replied:
will update

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -66,22 +65,81 @@ class ReplaceServerDetails extends Component {
+      data['role'] = this.props.data['role'];
+    }
+    else { // if it is non-compute node, only replace the mac-addr and ipmi info
+      data = Object.assign({}, this.props.data);
```
Again, we should not be feeding back data to the caller that it supplied to us.

GarySmith commented on src/pages/UpdateServers.js:
```diff
@@ -52,14 +54,15 @@ class UpdateServers extends BaseUpdateWizardPage {
 
       showSharedWarning: false,
 
-      serverToReplace: undefined,
```
FYI, a trailing comma in object lists is officially supported in JavaScript now, so you don't have to remove these, but I won't object if you do.

GarySmith commented on src/components/ReplaceServerDetails.js:
```diff
@@ -188,6 +248,33 @@ class ReplaceServerDetails extends Component {
     }
   }
 
+  isCompute = () => {
```
Please move this function into `ModelUtils`. It is clearly something that would be useful to have in several places in the code. It would be better to name it `isComputeNode` to make it clearer what we are talking about.

GarySmith commented on src/pages/UpdateServers.js:
```diff
@@ -115,16 +118,36 @@ class UpdateServers extends BaseUpdateWizardPage {
   }
 
-  updateServerForReplaceServer = (server) => {
+  getReplaceProps = (isCompute) => {
```
This should be calling `isComputeNode` from `ModelUtils`.

GarySmith commented on src/pages/UpdateServers.js:
```diff
@@ -115,16 +118,36 @@ class UpdateServers extends BaseUpdateWizardPage {
+        });
+    }
+    // for non-compute host replacement, if user added info manually, won't
+    // save to saved servers
```
Why aren't controller nodes stored there as well?

GarySmith commented on src/pages/UpdateServers.js:
```diff
@@ -240,8 +277,8 @@ class UpdateServers extends BaseUpdateWizardPage {
-  handleDoneReplaceServer = (server, wipeDisk, installOS) => {
-    this.replaceServer(server, wipeDisk, installOS);
+  handleDoneReplaceServer = (server, theProps) => {
+    this.replaceServer(server, theProps);
```
the old version was better because it was more explicit about what was really needed.

guglo replied:
the props will include extra params for UpdateServer to process...think I will keep all the extra stuffs in theProps especially when most of the process logic will be moved to UpdateServer..otherwise I need to pass wipe disk, installos, user/pass and selectedserverId...and some are not necessary there.

GarySmith commented on src/styles/components/replaceserverdetails.less:
```diff
@@ -25,10 +25,11 @@
+
   .replace-options {
     margin: 0.5em 0.5em 1.5em 1em;
+    &.more-bottom-margin {
```
this smells like a hack, but I'm not good enough with CSS to know how to improve it. Could this instead be handled by using an appropriate CSS selector like `:last-child` or `:last-of-type` or something?

guglo replied:
will look into it

guglo replied:
tried it with :last-child which is too generic...if use last-child...it applies to all that kind of last child in the dialog which I don't want that..I am not pro CSS like Jeremy, I will ask Jeremy tomorrow.

jeremy-moffitt replied:
If those dont work I dont think that the current implementation is all that bad, not worth spending more than 20-30 minutes on

guglo replied:
Tried it ..since I have various sections contain server-details-container...adding last-child and last-of-type will apply multiple of last-child or last-of-type. I only need add a little bit spaces before the buttons when all the sections are expanded. So will keep the hackish way for now.

GarySmith commented on src/styles/components/serverutils.less:
```diff
@@ -142,6 +142,9 @@
 .message-line {
   line-height: 2em;
   margin: 3em 0 1em 1em;
+  &.smaller-margin {
```
Ditto

guglo replied:
will look into it

GarySmith commented on src/utils/ModelUtils.js:
```diff
@@ -66,37 +67,23 @@ export function getAllOtherServerIds (model, autoServers, manualServers, theId)
-    if (idx !== -1) {
-      modelIds.splice(idx, 1);
-    }
+  if(!isEmpty(theId)) {
```
Nice improvement!

GarySmith commented on src/utils/ModelUtils.js:
```diff
@@ -66,37 +67,23 @@ export function getAllOtherServerIds (model, autoServers, manualServers, theId)
-      if (idx !== -1) {
-        autoServerIds.splice(idx, 1);
-      }
+    if(!isEmpty(theId)) {
```
Ditto

guglo commented on src/pages/UpdateServers.js:
```diff
@@ -152,26 +175,40 @@ class UpdateServers extends BaseUpdateWizardPage {
+    let model;
+
+    // compute host replacement has oldServer
+    let isCompute = theProps.oldServer;
```
yeah...it gets a little messy here. Will re-arrange some code to make a  better flow.

Commit:
Added comments

1. Moved data done logic from ReplaceServerDetails to UpdateServers
2. Fixed issue when save to servers when add compute
3. Added isCompute to ModelUntils so the usage is consistent
4. Minor styling code move

GarySmith reviewed (changes requested):
Thanks for the changes. I think it only needs the change of `isCompute` to `isComputeNode` or `hasComputeRole` for clarity

Commit:
Changed isCompute to isComputeNode
